src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs

```diff
@@ -6692,6 +6715,278 @@ public void Main()
+}
+", parseOptions: TestOptions.Regular8);
+
+            c.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void EnsuresTrue_Debug_Assert1_NotNullAndNotEmpty()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+class C
+{
+    void Main(string? c)
+    {
+        System.Diagnostics.Debug.Assert(c != null && c != """");
+        c.ToString();
+    }
+}
+", parseOptions: TestOptions.Regular8);
+
+            c.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void EnsuresTrue_Debug_Assert1_NotNullOrUnknown()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+class C
+{
+    void Main(string? c, bool b)
+    {
+        System.Diagnostics.Debug.Assert(c != null || b);
+        c.ToString();
+    }
+}
+", parseOptions: TestOptions.Regular8);
+
+            c.VerifyDiagnostics(
+                // (7,9): warning CS8602: Possible dereference of a null reference.
+                //         c.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c").WithLocation(7, 9)
+                );
+        }
+
+        [Fact]
+        public void EnsuresTrue_Debug_Assert2_NotNull()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+class C
+{
+    void Main(C? c)
+    {
+        System.Diagnostics.Debug.Assert(c != null, ""hello"");
+        c.ToString();
+    }
+}
+", parseOptions: TestOptions.Regular8);
+
+            c.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void EnsuresTrue_Debug_Assert3_NotNull()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+class C
+{
+    void Main(C? c)
+    {
+        System.Diagnostics.Debug.Assert(c != null, ""hello"", ""world"");
+        c.ToString();
+    }
+}
+", parseOptions: TestOptions.Regular8);
+
+            c.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void EnsuresTrue_Debug_Assert4_NotNull()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+class C
+{
+    void Main(C? c)
+    {
+        System.Diagnostics.Debug.Assert(c != null, ""hello"", ""world"", new object[] { });
+        c.ToString();
+    }
+}
+", parseOptions: TestOptions.Regular8);
+
+            c.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void EnsuresTrue_Debug_Assert_IsNull()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+class C
+{
+    void Main(C c)
+    {
+        System.Diagnostics.Debug.Assert(c == null, ""hello"");
+        c.ToString();
+    }
+}
+", parseOptions: TestOptions.Regular8);
+
+            c.VerifyDiagnostics(
+                // (6,41): hidden CS8606: Result of the comparison is possibly always false.
+                //         System.Diagnostics.Debug.Assert(c == null, "hello");
+                Diagnostic(ErrorCode.HDN_NullCheckIsProbablyAlwaysFalse, "c == null").WithLocation(6, 41)
+                );
+        }
+
+        [Fact]
+        public void EnsuresTrue_Debug_Assert_NoDuplicateDiagnostics()
```

Comment:
>_NoDuplicateDiagnostics [](start = 44, length = 23)

Consider changing this test and test below to not depend on `ExtraAnnotations`. That is, test with `[EnsuresTrue]` method declared in the test.